<launch>
  <!-- 1. Виртуальное время -->
  <param name="use_sim_time" value="true"/>

  <!-- 2. MAVROS SITL -->
  <include file="$(find mavros)/launch/apm.launch">
    <arg name="fcu_url" value="udp://:14550@127.0.0.1:14557"/>
  </include>

  <!-- 3. Статический трансформ map -> odom через python-скрипт -->
  <node pkg="basic_nav" type="map_to_odom.py" name="map_to_odom" output="screen"/>

  <!-- 3.1. Статический трансформ map -> odom через стандартный tf -->
  <node pkg="tf" type="static_transform_publisher" name="static_map_to_odom"
        args="0 0 0 0 0 0 map odom 100"
        launch-prefix="bash -c 'sleep 1; $0 $@'" />

  <!-- 4. Публикация target-позиций для дрона -->
  <node pkg="basic_nav" type="path_to_setpoint.py" name="path_to_setpoint" output="screen">
    <param name="waypoint_radius" value="0.5"/>
    <param name="altitude" value="1.5"/>
    <param name="use_custom_planner" value="true"/>
    <param name="takeoff_height" value="2.0"/>
  </node>

  <!-- 5. Загрузка URDF -->
  <param name="robot_description" command="$(find xacro)/xacro '$(find basic_nav)/urdf/drone_urdf.urdf'"/>
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="state_publisher" output="screen">
    <param name="publish_frequency" value="20.0"/>
  </node>

  <!-- 6. Статик base_link -> lidar_link -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_lidar"
        args="0 0 0.2 0 0 0 base_link lidar_link 100"
        launch-prefix="bash -c 'sleep 1; $0 $@'" />

  <!-- 7. Ретрансляция лидара с улучшенной производительностью -->
  <node pkg="topic_tools" type="relay" name="scan_relay"
        args="/drone1/scan /scan" output="screen">
    <param name="queue_size" value="20"/>
    <param name="unreliable" value="true"/>
  </node>

  <!-- 8. Map Server -->
  <node pkg="map_server" type="map_server" name="map_server"
        args="$(find basic_nav)/maps/my_map.yaml"/>

  <!-- 9. AMCL с улучшенными параметрами -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <param name="use_map_topic"   value="true"/>
    <param name="base_frame_id"   value="base_link"/>
    <param name="odom_frame_id"   value="odom"/>
    <param name="global_frame_id" value="map"/>
    <param name="transform_tolerance" value="0.5"/>
    <param name="update_min_d" value="0.1"/>
    <param name="update_min_a" value="0.1"/>
    <param name="min_particles" value="500"/>
    <param name="max_particles" value="5000"/>
    <param name="recovery_alpha_slow" value="0.0005"/>
    <param name="recovery_alpha_fast" value="0.1"/>
    <rosparam file="$(find basic_nav)/config/amcl_params.yaml" command="load"/>
  </node>

  <!-- 10. Move Base с улучшенными параметрами -->
  <node pkg="move_base" type="move_base" name="move_base" output="screen" respawn="true">
    <rosparam file="$(find basic_nav)/config/costmap_params.yaml" command="load"/>
    <rosparam file="$(find basic_nav)/config/base_local_planner.yaml" command="load"/>
    <rosparam file="$(find basic_nav)/config/recovery_behaviors.yaml" command="load"/>
    <param name="planner_frequency" value="5.0"/>
    <param name="controller_frequency" value="10.0"/>
    <param name="base_global_planner" value="navfn/NavfnROS"/>
    <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS"/>
    <param name="recovery_behavior_enabled" value="true"/>
    <param name="clearing_rotation_allowed" value="true"/>
    <param name="conservative_reset_dist" value="0.5"/>
    <param name="oscillation_timeout" value="10.0"/>
    <param name="oscillation_distance" value="0.5"/>
    <remap from="cmd_vel" to="/cmd_vel"/>
  </node>

  <!-- 11. Улучшенный глобальный планировщик -->
  <node pkg="basic_nav" type="global_planner.py" name="global_planner" output="screen">
    <param name="plan_frequency" value="2.0"/>
    <param name="altitude" value="1.5"/>
    <param name="use_markers" value="true"/>
  </node>

  <!-- 12. Улучшенный локальный планировщик -->
  <node pkg="basic_nav" type="local_planner.py" name="local_planner" output="screen">
    <param name="obstacle_threshold" value="1.0"/>
    <param name="safety_distance" value="0.6"/>
    <param name="max_linear_speed" value="0.5"/>
    <param name="max_angular_speed" value="0.5"/>
    <param name="look_ahead_distance" value="1.0"/>
  </node>

  <!-- 13. Координатная синхронизация -->
  <node pkg="basic_nav" type="coordinate_sync.py" name="coordinate_sync" output="screen"/>

  <!-- 14. RViz с улучшенной конфигурацией -->
  <node pkg="rviz" type="rviz" name="rviz"
        args="-d $(find basic_nav)/rviz/basic_nav.rviz"/>
</launch>
